### ISSUE TITLE
Question on stream resume

### ISSUE THREAD

**Issue by chelliwell** [infostellarinc/stellarcli#124]
Hi, could you explain the use of the 'resume' option please? What is the difference between _re_starting the CLI the same as the first time, compared with using the resume? Thanks.

**Comment by davidk81** [comment]
When client/CLI connects to the server (without "resume" option), it's analogous to 'tuning' in to a radio station; CLI starts receiving satellite data starting from the time of 'connection' (no prior data will be sent).

On the other hand, when client/CLI (re)connects to a stream with the "resume" option, server will instead start sending data from the last known data packet received by the client. This is possible because data sent by the satellite is first uploaded into StellarStation's cloud data pipeline which can hold the data for a few hours. Note however, resume should happen within a couple minutes because currently the "client resume state" is configured to expire quickly in the cloud data pipeline.

The reason "resume" does not happen by default is because StellarStation supports multiple simultaneous client downloads. Each connection is treated as if it is a new client unless explicitly specified.

**Comment by davidk81** [comment]
To put it simply, we advise to: 
- Establish a **new stream** a few minutes before pass AOS (connect without "resume" option)
- If the CLI disconnects for whatever reason **during** a pass, connect with **resume** option; the "stream-id" is provided when the stream is first created.

"stream-id" should be provided by the CLI when the `-v` (`--verbose`) flag is enabled.

**Comment by chelliwell** [comment]
Ah, I see, thanks.
So the restart command line can leave other params as-is, and just add "--stream-id _idstring_"?
I do though have a slight concern though that verbose is going to create a **lot** of output as well as the stream id.

**Comment by chelliwell** [comment]
Our downlink will be 25Mbps - we could disable capturing stdout once the stream-id has been parsed, but we might ultimately wish to leave it enabled so as to capture the stats output too.
Would it be possible to output the stream-id on the first packet received, when '-r' not used, and _not_ gated by '--verbose' ....?

**Comment by davidk81** [comment]
I agree we should be ensuring the stream-id is always displayed even when CLI `--verbose` is not turned on. @danada how do you feel about this?

**Comment by danada** [comment]
The client should try to reconnect by itself automatically if there are any disruptions. If reconnecting with `-r` and the stream ID, we would lose the flow-control last ACK ID, wouldn't we @davidk81 ?

With that said I do think it's a good idea to always display the stream ID when opening a stream.

**Comment by davidk81** [comment]
@danada yes, if the CLI gives up, the flow-control lask ACK ID (part of the "client resume state" would become out-of-sync with the server. if this happens, server may re-send data packets that has already been received by the client (hence the possible duplicate data). The even of 'lost' data should be very rare in theory once the data has made it to the cloud data pipeline; i believe all components of our data pipeline are built with redundancies.

I think we should also increase retry duration to minimize the chance of transient errors.

**Comment by davidk81** [comment]
to summarize, let't turn this thread into the following action item:

- make sure stream-id is always displayed on connection success, regardless of `--verbose` flag

(let's handle the retry issue on [issue-122](https://github.com/infostellarinc/stellarcli/issues/122))

**Comment by danada** [comment]
The change to always show the stream ID (without verbose flag) has been merged. We haven't scheduled the next release of the CLI, but in the meantime you should be able to build from source and use the CLI for testing.

I'm going to close this issue, but feel free to open it up (or a new one) if you have any more questions regarding the resume stream functionality.

**Comment by chelliwell** [comment]
Great, thanks. 
(The way our data block is built up, duplicate data wouldn't actually be a problem to me).

**Comment by chelliwell** [comment]
> The change to always show the stream ID (without verbose flag) has been merged. We haven't scheduled the next release of the CLI, but in the meantime you should be able to build from source and use the CLI for testing.
> 
> I'm going to close this issue, but feel free to open it up (or a new one) if you have any more questions regarding the resume stream functionality.

Would you expect the fakeserver to be generating a non-empty Stream ID?

**Comment by davidk81** [comment]
fake-server is indeed sending empty stream-id. the reason we are hesitant to implement stream-id is because it would be quite involved to make fake-server behave the the same way as the real server (considering features such as stream resume).

i've created a [branch where hard-coded stream-id is sent by fake-server;](https://github.com/davidk81/stellarstation-api/tree/fake-server-hardcode-stream-id) i hope you are able to run this. however please keep in mind this is only to help you do parsing tests with fake-server. If you plan to test stream-resume, i'd highly recommend testing against a test account. @danada what do you think?

**Comment by chelliwell** [comment]
Thanks for that :)
[I've been having some email conversation with Dan re. testing]

**Comment by chelliwell** [comment]
Can I just double-check a couple more details:
The stream-id is always one 'word' (no whitespace)?
The output message is sent to stderr? (!)

**Comment by davidk81** [comment]
- **stream-id** is always one word
- output message seem to be writing inconsistently to stdout and stderr at this time. i created a separate issue https://github.com/infostellarinc/stellarcli/issues/126
